### DISCUSSION TITLE
Analytical Jacobian or Geometrical Jacobian

### DISCUSSION THREAD

**Discussion by mahyaret** [stack-of-tasks/pinocchio#1649] [Q&A] [answered]
Hello,
Thanks for the great kinodynamics library. I was wondering which functions for computing Jacobian addresses Analytical Jacobian and which one is computing Geometrical Jacobian?
Thanks,
Mahyar

**Comment by jmirabel** [comment]
What is a *Geometrical Jacobian* ?

**Reply by mahyaret** [reply]
Geometrical Jacobian is when the Jacobian is computed according to a geometric technique in which the contributions of each joint velocity to the components of end-effector linear and angular velocity are determined. Analytical Jacobian is when Jacobian is determined via differentiation of the forward kinematics function with respect to the joint variables. 
more details can be found here: [https://www.etedal.net/2014/12/analytical-jacobian-vs-geometrical.html](https://www.etedal.net/2014/12/analytical-jacobian-vs-geometrical.html)

**Comment by jcarpent** [comment, marked as answer]
You can use `pin.computeJointJacobians(model, data, q)` and `pin.getJointJacobian(model, data, joint_id, pin.LOCAL)`.
This will corresponds to kinematic Jacobian associated to the joint velocity expressed in the LOCAL frame of the joint.

**Reply by mahyaret** [reply]
Based on the example code provided for iterative inverse kinematic [here](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/md_doc_b-examples_i-inverse-kinematics.html), I thought `pin.computeJointJacobian(model, data,q)` is analytical Jacobian. I am wondering if there are separate functions for Geometrical Jacobian and Analytical Jacobian.

**Reply by jcarpent** [reply]
What do you want to do?

**Reply by jcarpent** [reply]
To be more precise, in the context of IK, analytical and geometric Jacobians match as the quantity being optimized being the spatial velocity vector attached to the joint.

**Reply by mahyaret** [reply]
I just wanted to be sure which implementation I am using when I call `pin.computeJointJacobian(model, data, q)`. Is it Analytical or Geometrical? since they are not the same thing and related using: J_{Analytical} = T J_{Geometrical}, where T is a transformation matrix. For instance, it is going to be important for mapping wrench vector to joint torques, \tau = J^T F

**Reply by jcarpent** [reply]
In Pinocchio, the kinematic Jacobians map joint velocities to frame spatial velocity. Then, `T` can be any transformation between spatial velocity vector.

**Reply by mahyaret** [reply]
so for instance, if we use `pin.computeJointJacobian(model, data, q)` for \dot x = J \dot q then \dot x  is [\dot x, \dot y, \dot z, \omega x, \omega y, \omega z]^T where \omega is the angular velocity, am I right?

**Reply by jcarpent** [reply]
Yes

**Reply by mahyaret** [reply]
Thanks!

**Comment by LHT-Liu** [comment]
import pinocchio as pin

from pinocchio.visualize import MeshcatVisualizer

urdf_path = "../urdf/Franka_urdf/fr3_without_hand.urdf"

model: pin.Model = pin.buildModelFromUrdf(urdf_path)

data = model.createData()

q0 = pin.neutral(model)

end_frame = model.getJointId("fr3_joint8")

pin.computeJointJacobians(model, data, q0)

end_jacobian = pin.getJointJacobian(model, data, end_frame, pin.LOCAL)




why did the following error occur? The IDE I used is pycharm and the system is macbook.

![image](https://github.com/stack-of-tasks/pinocchio/assets/64388319/009841de-30c0-4e88-aa93-66a6c991dc7b)

**Reply by fabinsch** [reply]
see https://github.com/stack-of-tasks/pinocchio/issues/2181. Please avoid cross-posting the same issue in different places.
